### ISSUE TITLE
Support wildcards for coverage report paths

### ISSUE THREAD

**Issue by vilchik-elena** [SonarSource/SonarJS#578]
Currently property for coverage LCOV report paths doesn't accept paths with wildcards. This might be very useful when path is not stable (e.g. see [question on SOF](http://stackoverflow.com/questions/43319073/wildcards-for-lcov-info-path-in-the-sonarqube-scanners)).

Property name: **sonar.javascript.lcov.reportPaths**

* [Docs for coverage import in SonarJS](https://docs.sonarqube.org/display/PLUG/JavaScript+Coverage+Results+Import)
* [Docs for patterns in SonarQube](https://docs.sonarqube.org/display/SONAR/Narrowing+the+Focus#NarrowingtheFocus-patterns)

We should decide whether this feature should be implemented for old properties (still supported for SQ < 6.2):
- **sonar.javascript.lcov.reportPath**
- **sonar.javascript.lcov.itReportPath**

**Example:**
_"sonar.javascript.lcov.reportPaths=reports/*"_ would match 
_reports/report1.lcov
reports/report2.lcov_

_"sonar.javascript.lcov.reportPaths=**/report.lcov"_ would match 
_reports1/report.lcov
reports2/report.lcov_

**Comment by ivandalbosco** [comment]
+1
As regards SQ < 6.2: I suggest we simply pick the cheapest solution between supporting and non supporting.

**Comment by ideadapt** [comment]
I use something like

```JS
{ type: 'lcov', dir: 'test/reports/unit/coverage/lcov/', subdir: function(browser) {
    // keep a consistent browser name across different OS & browser versions
    return browser.toLowerCase().split(/[ /-]/)[0];
  }
}
```

to generate a version neutral path. Found at https://github.com/karma-runner/karma-coverage/blob/master/docs/configuration.md

**Comment by vilchik-elena** [comment]
There is implementation for wildcard support in sonarGo and SonarPython. We should migrate it to sonar-analyzer-commons

**Comment by hzdavid** [comment]
so Could we use wildcards to . sonar.jacoco.reportPaths now ?

**Comment by fadc80** [comment]
Is there any chance to see this issue resolved soon?

**Comment by roy-bakker-new10** [comment]
Would still love this!

**Comment by dblees** [comment]
So until this is supported this feels brittle. every time a codebase changes, this forces maintenance of configuration to capture reports and classes. please consider this feature. thanks.

**Comment by Rai-Rai** [comment]
👍 
Need support for wildcards for `sonar.typescript.lcov.reportPaths` really desperate....

**Comment by vilchik-elena** [comment]
@Rai-Rai since last release (SonarTS 2.0 and SonarJS 6.0) there is single logic and property to load coverage for TS and JS: sonar.javascript.lcov.reportPaths

**Comment by brenuart** [comment]
Would love this too - and not only for SonarJS but not only for coverage report paths but others as well, like:
- sonar.javascript.lcov.reportPaths
- sonar.typescript.tslint.reportPaths
- sonar.testExecutionReportPaths (generic report)

As @vilchik-elena said, they all share a common logic located in `sonar-analyzer-commons` - can't we simply adapt it to support ANT-style patterns for instance?

Would you consider a PR for this ?

**Comment by brenuart** [comment]
Any chance this issue gets some attention from official SonarQube guys? At least some indication about whether this feature request makes sense or not.. ?

**Comment by pburkindine** [comment]
Also badly need this feature for js/ts

**Comment by clok** [comment]
Bump. This would be very helpful!

**Comment by lpcruz** [comment]
Bumping this as well--- currently in a situation where we need to support multiple lcov reports so having it be a capability of Sonar itself would be amazing.

**Comment by mariamatthews** [comment]
Hi guys, could I suggest using a quick bash script before your SonarQube runs that would identify the file/folder locations for you as a workaround?

Putting the below path and updating the `packages` folder name to the root folder of your projects should find the exact locations for you.

Code:
```bash
        
  - bash: |
      for entry in packages/*/
      do
      if [ "$entry" == "packages/common/" ] ; then
            continue;
      fi
      testPaths=${testPaths}${entry}test-report.xml,
      coveragePaths=${coveragePaths}${entry}coverage/lcov.info,
      done
      testPaths=${testPaths%,}
      coveragePaths=${coveragePaths%,}
      echo "Test and coverage reports paths found: "
      echo "Test reports: $testPaths"
      echo "Coverage reports: $coveragePaths"
      echo "Setting pipeline variables..."
      echo "##vso[task.setvariable variable=TESTPATHS]$testPaths"
      echo "##vso[task.setvariable variable=COVERAGEPATHS]$coveragePaths"
    displayName: Getting report paths for SonarQube
```

The above task would iterate over the folders and add the files found within the temporary Azure Pipelines variables.
The `if` statement is where you could add any folders you would like it to skip.

This means that you can set SonarQube `reportPaths` like:

```bash
      sonar.testExecutionReportPaths=$(TESTPATHS)
      sonar.javascript.lcov.reportPaths=$(COVERAGEPATHS)
```

Let me know if you have found it useful and if it worked for you :orange_heart:

**Comment by dclowd9901** [comment]
I am finding that wildcards for reportPaths doesn't actually work:

This setting doesn't work:
```
sonar.javascript.lcov.reportPaths=lcov/lcov*.info
```

This setting does:
```
sonar.javascript.lcov.reportPaths=lcov/lcov-1.info,lcov/lcov-2.info,lcov/lcov-3.info,lcov/lcov-4.info,lcov/lcov-5.info,lcov/lcov-6.info,lcov/lcov-7.info,lcov/lcov-8.info,lcov/lcov-9.info,lcov/lcov-10.info,lcov/lcov-11.info,lcov/lcov-12.info,lcov/lcov-13.info,lcov/lcov-14.info,lcov/lcov-15.info

**Comment by hoch4** [comment]
This workaround works well too

```shell
#!/bin/sh
coverage=$(find -type f -name '*coverage.lcov' -printf "%p,")
testResults=$(find -type f -name '*test-result.xml' -printf "%p,")

mvn sonar:sonar -o \
  -Dsonar.projectKey=com.my.project.key.from.sonarqube \
  -Dsonar.host.url=http://mysonarqubeserver.com:9000 \
  -Dsonar.login=kd94kaek94ad9duiope67fbn89vbnnn89dw \
  -Dsonar.javascript.lcov.reportPaths="${coverage}" \
  -Dsonar.testExecutionReportPaths="${testResults}"
```

**Comment by arobinson** [comment]
I'm still having this problem. I've tried both * and ** with no luck with `sonar.javascript.lcov.reportPaths`

Using
```
export SONAR_SCANNER_VERSION="5.0.1.3006"
```
And
```
"sonarqube-scanner": "^3.3.0
```

Error:
```
14:47:08  INFO: No LCOV files were found using /home/ec2-user/workspace/Client/some-workspace/test-results/coverage/*.lcov.info
14:47:08  WARN: No coverage information will be saved because all LCOV files cannot be found.
```
I archived the `test-results/coverage` folder as part of the Jenkins job, and verified the page is correct. Also, from the error, it does not sound like it is using the code that was in the fix for this issue.

What version needs to be used to obtain the fix?
